feat(allocator): remove drop operations from Vec2

Dunqing opened the pull request:
Just like #6623. Since we have control over the `Vec` implementation, we can remove the `drop` operations and eventually revert the changes made in #6623 directly.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9679?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9679** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9679?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9679?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9656** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9656?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9655** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9655?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9647** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9646** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9646?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9679).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
> I am not sure which `drop` can be removed, I may need to take a look at Bumpalo's `Bump` implementation.

We can remove loads. But I'd suggest starting by just removing:

* `impl<'bump, T> Drop for Vec<'bump, T>`
* `impl<'a, T> Drop for RawVec<'a, T>`

That'd bring `vec2::Vec` up to parity with current `Vec`. It'd be enough to then be able to remove `ManuallyDrop` from `Vec` so `Vec` becomes just `struct Vec(vec2::Vec);`.

Removing further `Drop`-related code would be an improvement over what we have now. I'd be interested to see if it affects benchmarks as we take them out one by one.

But we may want to simplify/remove other code first. `bumpalo`'s implementation offers various options which we don't use, and they clutter the code and make it harder to see what's going on. I suggest we discuss next steps once we've got the first parts merged.

overlookmotel reviewed (commented):
Could we limit the scope of this PR to just removing these 2?

* `impl<'bump, T> Drop for Vec<'bump, T>`
* `impl<'a, T> Drop for RawVec<'a, T>`

Those 2 `Drop` impls are definitely fine to remove. The other ones are less clear without reading all the code around them.

I think it's preferable to go step-by-step on this. It'd be very easy to make a mistake somewhere and cause UB, so I think it's better to err on the side of caution.

Dunqing commented:
> Could we limit the scope of this PR to just removing these 2?
> 
> * `impl<'bump, T> Drop for Vec<'bump, T>`
> * `impl<'a, T> Drop for RawVec<'a, T>`
> 
> Those 2 `Drop` impls are definitely fine to remove. The other ones are less clear without reading all the code around them.
> 
> I think it's preferable to go step-by-step on this. It'd be very easy to make a mistake somewhere and cause UB, so I think it's better to err on the side of caution.

This PR is already just removing these 2 which you suggested, additionally, removed `mem::forgot` because non-drop calling `mem::forgot` is meaningless.

Dunqing commented:
> But could you replace them with comments saying something like this?
> 
> ```rust
> // Don't need `mem::forget(self)` here, because `Vec` does not implement `Drop`
> ```


Done!

overlookmotel commented:
### Merge activity

* **Mar 14, 3:50 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
feat(allocator): remove drop operations from Vec2 (#9679)

Just like #6623. Since we have control over the `Vec` implementation, we can remove the `drop` operations and eventually revert the changes made in #6623 directly.
